### ISSUE TITLE
Migration fails when removing the first child of a node

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3346]
When using the migrateX_X_X methods to perform migration, if the first child of the element is removed then the migration will fail with an UnknownFieldException. 

Take the following XML configuration. 

`
<updateSettings>
    <firstLogin>false</firstLogin>
    <statsEnabled>false</statsEnabled>
    <updatesEnabled>true</updatesEnabled>
    <updateUrl>http://updates.mirthcorp.com</updateUrl>
</updateSettings>
`

If firstLogin is removed via element.removeChild("firstLogin"), then when unmarshal is called this will fail. However if the XML is modified such that firstLogin is no longer the first child of updateSettings, then the migration is successful.

`
<updateSettings>
    <statsEnabled>false</statsEnabled>
    <firstLogin>false</firstLogin>
    <updatesEnabled>true</updatesEnabled>
    <updateUrl>http://updates.mirthcorp.com</updateUrl>
</updateSettings>
`

In the XStream class AbstractReflectionConverter at line 310, the following code is called

`
reader.moveDown();

String originalNodeName = reader.getNodeName();
`

Even though firstLogin was removed in migrateX_X_X, the value for originalNodeName is still firstLogin. This will then fail at line 342 because the result of the following code is null, which causes the exception to be thrown.

`type = mapper.realClass(originalNodeName);`

I think the underlying problem is that we update the node directly but not the reader. Requires some more research at this point.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3446
Reporter: wayneh
Created: 2014-09-09T19:18:16.000-0700

**Comment by rbeckman-nextgen** [comment]
XStream's DomReader keeps track of the child elements in a separate List, and that list isn't reloaded unless the depth is explicitly changed. XStream isn't expecting the underlying element to be dynamically modified like we're doing, so whenever we do that, we need to manually update the DomReader's child list.

Imported Comment. Original Details:
Author: narupley
Created: 2014-09-10T11:42:44.000-0700

**Comment by rbeckman-nextgen** [comment]
Fixed in revision 7354. Whenever migration is performed, the DomReader is notified to update its list of child elements.

Imported Comment. Original Details:
Author: narupley
Created: 2014-09-10T12:32:27.000-0700

**Comment by rbeckman-nextgen** [comment]
Was able to repro this issue when implementing my fix for MIRTH-3372. UpdateSettings would not successfully migrate when restoring a server config from version 2 and version 3. 

I've verified that server configs from previous versions now restore successfully. I've also made channels with all source and destination types from versions 2 and 3 and was able to successfully migrate them to 3.1.

Imported Comment. Original Details:
Author: leilanim
Created: 2014-09-12T19:34:59.000-0700
